crates/matrix-sdk-base/src/sliding_sync.rs

```diff
@@ -30,60 +30,73 @@ impl BaseClient {
             // next_batch,
             rooms,
             lists,
+            extensions,
             // FIXME: missing compared to v3::Response
             //presence,
-            //account_data,
-            //to_device,
-            //device_lists,
-            //device_one_time_keys_count,
-            //device_unused_fallback_key_types,
             ..
         } = response;
 
-        // FIXME not yet supported by sliding sync. see
-        // https://github.com/matrix-org/matrix-rust-sdk/issues/1014
-        // #[cfg(feature = "encryption")]
-        // let to_device = {
-        //     if let Some(o) = self.olm_machine().await {
-        //         // Let the crypto machine handle the sync response, this
-        //         // decrypts to-device events, but leaves room events alone.
-        //         // This makes sure that we have the decryption keys for the room
-        //         // events at hand.
-        //         o.receive_sync_changes(
-        //             to_device,
-        //             &device_lists,
-        //             &device_one_time_keys_count,
-        //             device_unused_fallback_key_types.as_deref(),
-        //         )
-        //         .await?
-        //     } else {
-        //         to_device
-        //     }
-        // };
-
-        if rooms.is_empty() {
-            // nothing for us to handle here
+        if rooms.is_empty() && extensions.is_empty() {
+            // we received a room reshuffling event only, there won't be anything for us to
+            // process. stop early
             return Ok(SyncResponse::default());
         };
 
+        let v4::Extensions { to_device, e2ee, account_data, .. } = extensions;
+
+        let to_device_events = to_device.map(|v4| v4.events).unwrap_or_default();
+
+        #[cfg(feature = "e2e-encryption")]
+        let to_device_events = {
+            if let Some(e2ee) = &e2ee {
+                if let Some(o) = self.olm_machine() {
+                    // Let the crypto machine handle the sync response, this
+                    // decrypts to-device events, but leaves room events alone.
+                    // This makes sure that we have the decryption keys for the room
+                    // events at hand.
+                    o.receive_sync_changes(
+                        to_device_events,
+                        &e2ee.device_lists,
+                        &e2ee.device_one_time_keys_count,
+                        e2ee.device_unused_fallback_key_types.as_deref(),
+                    )
+                    .await?
+                } else {
+                    to_device_events
+                }
+            } else {
+                to_device_events
+            }
+        };
```

Comment:
I see the dilemma you had when writing this code. Maybe by using `Option::map`, we can simplify this code a little bit (not tested, maybe there is some lifetime issues, not sure):

```suggestion
        let to_device_events = e2ee
            .zip(self.olm_machine())
            .map(|(e2ee, machine)| {
                machine
                    .receive_sync_changes(
                        to_device_events,
                        &e2ee.device_lists,
                        &e2ee.device_one_time_keys_count,
                        e2ee.device_unused_fallback_key_types.as_deref(),
                    )
                    .await
            })
            .transpose()?
            .unwrap_or(to_device_events);
```

Reply:
/me wants if-let-chains!

Reply:
@Hywan your suggestion doesn't work because you can't use `.await` inside a regular closure. One level of nesting could be removed by writing `if let Some((e2ee, o)) = e2ee.as_ref().and_then(|e2ee| Some((e2ee, self.olm_machine()?)))`, but it's not that nice. Or `if let (Some(e2ee), Some(o)) = (&e2ee, self.olm_machine())`, but that has different semantics in that it will call `olm_machine()` regardless of whether `e2ee` is `Some` (maybe that's fine).

Reply:
So let's call `.await` outside. I had something like that in mind:

```rs
        let to_device_events = match e2ee.as_ref().zip(self.olm_machine()).map(|(e2ee, machine)| {
            machine.receive_sync_changes(
                to_device_events,
                &e2ee.device_lists,
                &e2ee.device_one_time_keys_count,
                e2ee.device_unused_fallback_key_types.as_deref(),
            )
        }) {
            Some(to_device_events) => to_device_events.await?,
            None => to_device_events,
        };
```

But in this case, `to_device_events` is moved inside the the closure, so it cannot be used in the last `None => to_device_events` line. The problem can be simply solved by cloning the `Vec`, but we don't want that.

Let's keep the code as it is now then :-).